draft-west-cookie-incrementalism.md

```diff
@@ -401,85 +404,275 @@ from "have the same registrable domain" (or "is an exact match for") to say
+       and ignore the newly created cookie entirely.
+
+    2. Update the creation-time of the newly-created cookie to match
+       the creation-time of the old-cookie.
+
+    3. Remove the old-cookie from the cookie store.
+    (Notice that this algorithm maintains the invariant that there
+    is at most one such cookie.)
 ~~~
 
-Second, alter The Cookie Header algorithm defined in Section 5.5 of {{RFC6265bis}} to take the
-`scheme` into account when deciding which cookies to deliver by adding another condition to the
-list in Step 1 of the algorithm:
+With cookie storage taken care of now we move onto sending cookies. In
+section 5.6.1 replace step 1 with the following:
 
 ~~~
-*   The cookies' `scheme` matches the scheme component of request-uri's origin.
+1. Let initial-cookie-list be the set of cookies from the cookie store that
+    meets all of the following requirements:
+
+    * Either:
+
+      - The cookie's host-only-flag is true, and the canonicalized
+        request-host is identical to the cookie's domain.
+
+    Or:
+
+      - The cookie's host-only-flag is false, and the canonicalized
+        request-host domain-matches the cookie's domain.
+
+    * The request-uri's origin's port component port-matches the
+      cookie.
+
+    * The request-uri's origin's scheme component is identical to the
+      cookie's scheme
 ~~~
 
-This seems like the minimal set of changes necessary. We could do other cleanup, including removing
-the `Secure` attribute, as this mechanism obviates it entirely, altering the eviction algorithm to
-prefer discarding non-secure schemes, etc.
+This steps collects all cookies that could potentially be sent on this
+request. Next add a new step that filters this list down to the cookies
+that should actually be sent.
+
+~~~
+2. Let cookie-list be a new empty list.
+   For each cookie in initial-cookie-list:
+   1. Continue to the next cookie if cookie's host-only-flag is false
+      and initial-cookie-list contains any cookies that meet all of
+      the following criteria:
+        * their host-only-flag is true.
+        * their name matches cookie's name.
+   2. Add cookie to cookie-list if it meets all of the following
+      requirements:
+        * The retrieval's URI's path path-matches the cookie's path.
+        
+        * If the cookie's http-only-flag is true, then exclude the
+          cookie if the retrieval's type is "non-HTTP".
+          
+        * If the cookie's same-site-flag is not "None" and the retrieval's same-site
+          status is "cross-site", then exclude the cookie unless all of the
+          following conditions are met:
+
+          * The retrieval's type is "HTTP".
+          * The same-site-flag is "Lax" or "Default".
+          * The HTTP request associated with the retrieval uses a "safe" method.
+          * The target browsing context of the HTTP request associated with the
+            retrieval is a top-level browsing context.
```

Comment:
This feels clunky, but I wasn't sure how to better describe "Don't send an domain cookie if there is a matching origin cookie anywhere on the origin"